src/simulation/elements/LSNS.cpp

```diff
@@ -52,36 +52,104 @@ int Element_LSNS::update(UPDATE_FUNC_ARGS)
 	if (parts[i].life)
 	{
 		parts[i].life = 0;
-		for (rx=-2; rx<3; rx++)
-			for (ry=-2; ry<3; ry++)
+		for (rx = -2; rx < 3; rx++)
+			for (ry = -2; ry < 3; ry++)
 				if (BOUNDS_CHECK && (rx || ry))
 				{
-					r = pmap[y+ry][x+rx];
+					r = pmap[y + ry][x + rx];
 					if (!r)
 						continue;
 					rt = TYP(r);
-					if (sim->parts_avg(i,ID(r),PT_INSL) != PT_INSL)
+					if (sim->parts_avg(i, ID(r), PT_INSL) != PT_INSL)
 					{
-						if ((sim->elements[rt].Properties&PROP_CONDUCTS) && !(rt==PT_WATR||rt==PT_SLTW||rt==PT_NTCT||rt==PT_PTCT||rt==PT_INWR) && parts[ID(r)].life==0)
+						if ((sim->elements[rt].Properties&PROP_CONDUCTS) && !(rt == PT_WATR || rt == PT_SLTW || rt == PT_NTCT || rt == PT_PTCT || rt == PT_INWR) && parts[ID(r)].life == 0)
 						{
 							parts[ID(r)].life = 4;
 							parts[ID(r)].ctype = rt;
-							sim->part_change_type(ID(r),x+rx,y+ry,PT_SPRK);
+							sim->part_change_type(ID(r), x + rx, y + ry, PT_SPRK);
 						}
 					}
 				}
 	}
-	for (rx=-rd; rx<rd+1; rx++)
-		for (ry=-rd; ry<rd+1; ry++)
-			if (x+rx>=0 && y+ry>=0 && x+rx<XRES && y+ry<YRES && (rx || ry))
+	for (rx = -rd; rx < rd + 1; rx++)
+		for (ry = -rd; ry < rd + 1; ry++)
+			if (x + rx >= 0 && y + ry >= 0 && x + rx < XRES && y + ry < YRES && (rx || ry))
 			{
-				r = pmap[y+ry][x+rx];
-				if(!r)
-					r = sim->photons[y+ry][x+rx];
-				if(!r)
+				r = pmap[y + ry][x + rx];
+				if (!r)
+					r = sim->photons[y + ry][x + rx];
+				if (!r)
 					continue;
-				if (parts[ID(r)].life > parts[i].temp-273.15)
+				if (parts[ID(r)].life > parts[i].temp - 273.15)
 					parts[i].life = 1;
+			}                                                                                   //.Life serialization.(.tmp2 = turns serialization mode on)
+	bool setFilt = false;
+	int photonWl = 0;
+	for (int rx = -rd; rx <= rd; rx++)
+		for (int ry = -rd; ry <= rd; ry++)
+			if (x + rx >= 0 && y + ry >= 0 && x + rx < XRES && y + ry < YRES && (rx || ry))
+			{
+				int r = pmap[y + ry][x + rx];
+				if (!r)
+					r = sim->photons[y + ry][x + rx];
+				if (!r)
+					continue;
+				if (parts[i].tmp == 1 && TYP(r) != PT_LSNS && TYP(r) != PT_FILT)
+				{
+					setFilt = true;
+					photonWl = parts[ID(r)].life;
+				}
+			}
+	if (setFilt)
+	{
+		int nx, ny;
+		for (int rx = -1; rx <= 1; rx++)
+			for (int ry = -1; ry <= 1; ry++)
+				if (BOUNDS_CHECK && (rx || ry))
+				{
+					int r = pmap[y + ry][x + rx];
+					if (!r)
+						continue;
+					nx = x + rx;
+					ny = y + ry;
+					while (TYP(r) == PT_FILT)
+					{
+						parts[ID(r)].ctype = 0x10000000 + photonWl;
+						nx += rx;
+						ny += ry;
+						if (nx < 0 || ny < 0 || nx >= XRES || ny >= YRES)
+							break;
+						r = pmap[ny][nx];
+					}
+				}
+	}
+	// .Life Deserialization (.tmp2 = turns deserialization mode on).Sets nearby particle's life.
+	bool deserializelife = parts[i].tmp == 2;
```

Comment:
When this is false, the loop below doesn't need to run. It would be more efficient if you put an if (parts[i].tmp == 2) around this whole loop (replacing the deserializelife checks in the loop below)

Reply:
This is still not resolved. I said you should have the if around the entire loop, because the loop doesn't need to run at all unless .tmp is 2. It will lag the game always running the loop